hawkbit-repository/hawkbit-repository-jpa/src/test/java/org/eclipse/hawkbit/repository/jpa/ControllerManagementTest.java

```diff
@@ -989,4 +1001,154 @@ public void createActionStatusWithTooManyMessages() {
+        final Long actionId = createAndAssignDsAsDownloadOnly(TestdataFactory.DEFAULT_CONTROLLER_ID);
+        assertThat(actionId).isNotNull();
+        IntStream.range(0, 3).forEach(i -> controllerManagement
+                .addUpdateActionStatus(entityFactory.actionStatus().create(actionId).status(Status.DOWNLOADED)));
+
+        assertActionStatus(actionId, TestdataFactory.DEFAULT_CONTROLLER_ID, TargetUpdateStatus.IN_SYNC,
+                Status.DOWNLOADED, Status.DOWNLOADED, false);
+
+        assertThat(actionStatusRepository.count()).isEqualTo(4);
+        assertThat(controllerManagement.findActionStatusByAction(PAGE, actionId).getNumberOfElements()).isEqualTo(4);
+        assertThat(actionRepository.activeActionExistsForControllerId(TestdataFactory.DEFAULT_CONTROLLER_ID))
+                .isEqualTo(false);
+    }
+
+    @Test(expected = QuotaExceededException.class)
+    @Description("Verifies that quota is asserted when a controller reports too many DOWNLOADED events for a " +
+            "DOWNLOAD_ONLY action.")
+    public void quotaExceptionWhencontrollerReportsTooManyDownloadedMessagesForDownloadOnlyAction() {
+        final int maxMessages = quotaManagement.getMaxMessagesPerActionStatus();
+        testdataFactory.createTarget();
+        final Long actionId = createAndAssignDsAsDownloadOnly(TestdataFactory.DEFAULT_CONTROLLER_ID);
+        assertThat(actionId).isNotNull();
+
+        IntStream.range(0, maxMessages).forEach(i -> controllerManagement
+                .addUpdateActionStatus(entityFactory.actionStatus().create(actionId).status(Status.DOWNLOADED)));
+    }
+
+    @Test
+    @Description("Verifies that no quota exceptions are thrown when a controller reports too many FINISHED events for "
+            + "a DOWNLOAD_ONLY action.")
+    public void noQuotaExceptionWhenControllerReportsTooManyFinishedMessagesForDownloadOnlyAction() {
+        final int maxMessages = quotaManagement.getMaxMessagesPerActionStatus();
+        testdataFactory.createTarget();
+        final Long actionId = createAndAssignDsAsDownloadOnly(TestdataFactory.DEFAULT_CONTROLLER_ID);
+        assertThat(actionId).isNotNull();
+
+        try {
+            IntStream.range(0, maxMessages).forEach(i -> controllerManagement
+                    .addUpdateActionStatus(entityFactory.actionStatus().create(actionId).status(Status.FINISHED)));
+        } catch (QuotaExceededException e) {
+            fail("No QuotaExceededException should be thrown in this context!");
+        }
+
+        assertActionStatus(actionId, TestdataFactory.DEFAULT_CONTROLLER_ID, TargetUpdateStatus.IN_SYNC,
+                Status.FINISHED, Status.FINISHED, false);
+
+        assertThat(actionStatusRepository.count()).isEqualTo(2);
+    }
+
+    @Test
+    @Description("Verifies that a controller can report a FINISHED event for a DOWNLOAD_ONLY action.")
+    public void controllerReportsFinishedForOldDownloadOnlyActionAfterSuccessfulForcedAssignment() {
```

Comment:
This test is very long. Instead of code blocks with a comment it might be better to put the code blocks into 'Step'-methods